File: 백준/Silver/9012. 괄호/괄호.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int t = Integer.parseInt(br.readLine());

        for (int i = 0; i < t; i++) {
            String s = br.readLine();
            int cnt = 0;
            for (int j = 0; j < s.length(); j++) {
                char c = s.charAt(j);
                if (c == '(') {
                    ++cnt;
                } else if (c == ')') {
                    --cnt;
                    if (cnt < 0) {
                        break;
                    }
                }
            }
            if (cnt == 0) {
                System.out.println("YES");
            } else {
                System.out.println("NO");
            }
        }
        
    }

}

File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline

t = int(input())
for _ in range(t):
    s = input().strip()
    cnt = 0
    for i in range(len(s)):
        if s[i] == '(':
            cnt += 1
        else:
            cnt -= 1
            if cnt < 0:
                break
    if cnt == 0:
        print("YES")
    else:
        print("NO")
